### PR TITLE
[sql] Add support for omitting columns in INSERT SELECT

### LINKED ISSUE #165: Cannot insert using select if using partial column selection
Given
```
CREATE TABLE pk_single_row_hist (
  id INTEGER PRIMARY KEY,
  t TEXT DEFAULT CURRENT_TIMESTAMP,
  client INTEGER,
  value INTEGER
);
```

When
```
INSERT INTO pk_single_row_hist (client, value) SELECT client, value FROM pk_single_row;
```

Then
```
Error: line 2, character 7:
Incompatible select size
```

### PR DESCRIPTION
**Pull request by jberdine**
Values for columns that are either PRIMARY KEYs, NULLABLE, or have a
DEFAULT may be omitted when inserting rows. This PR adds support for
such omissions when INSERTing the rows returned by a SELECT.

The `computeInsert` function is refactored to expose
`computeInsertRow`, which contains the functionality to insert omitted
defaults, etc. This `computeInsertRow` function is then also called
when evaluating INSERT SELECT, in the mapper over the rows returned by
the query that inserts them into the table. Also, the error checking
code is adjusted accordingly.

closes #165

### CODE REVIEW

File: sql/src/SqlEval.sk [hunk 1]
```diff
@@ -149,75 +149,84 @@ fun computeInsert(
...
-      for (userDefinedCol in columnParams) {
-        if (!missingCols.containsKey(userDefinedCol)) {
-          error(pos, "Column not found: " + userDefinedCol);
-        };
-        missingCols.remove(userDefinedCol);
+    computeInsertRow(context, pos, columnParamsOpt, values, table, repeat)
+  })
+}
+
+fun computeInsertRow(
```
Comment (jberdine) [inline sql/src/SqlEval.sk:156]: Github does not show the diff well, but note that the body of this function is simply moved unchanged. No need to carefully compare.

File: sql/src/SqlEval.sk [hunk 2]
```diff
@@ -591,7 +631,23 @@ class Evaluator{options: Options, user: ?UserFile} {
           sinkDirName,
           (context, _writer, key, valueIter) ~> {
             src = SKStore.Path(selectDir.dirName, key);
-            entries = Vector[(key, (src, src, valueIter.collect(Array)))];
+            rows: Array<SKStore.File> = valueIter
+              .map(row -> {
+                row match {
+                | RowValues(cvalues, repeat) ->
```
Comment (jberdine) [inline sql/src/SqlEval.sk:637]: I have not found a way to write this without an unsafe down-cast like this. Is there a way?

File: sql/src/SqlEval.sk [hunk 3]
```diff
@@ -591,7 +631,23 @@ class Evaluator{options: Options, user: ?UserFile} {
           sinkDirName,
           (context, _writer, key, valueIter) ~> {
             src = SKStore.Path(selectDir.dirName, key);
-            entries = Vector[(key, (src, src, valueIter.collect(Array)))];
+            rows: Array<SKStore.File> = valueIter
```
Comment (jberdine) [inline sql/src/SqlEval.sk:634]: Typing fails without this annotation, but I haven't been able to find a better way to write this.

File: sql/src/SqlEval.sk (hunk 2 above)
Comment (gregsexton) [reply sql/src/SqlEval.sk:637]: Yeah, we have code like that everywhere. SKStore has some type safety but it's a layer on top which isn't always used.

File: sql/src/SqlEval.sk (hunk 1 above)
Comment (gregsexton) [reply sql/src/SqlEval.sk:156]: Thanks. Very helpful. :)

**Comment by gregsexton** [comment]
Looks great! Thanks for fixing!

I did just create a new issue (#175) if you're on a roll. ;) :D

**Pull request merged**